python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -171,274 +200,122 @@ def codegen(context, builder, sig, args):
-    def size(self):
-        return self.ntype.bitwidth // 8
-
-    @property
-    def alignment(self):
-        return self.size
-
-
-class TransformIteratorImpl:
-    def __init__(
-        self,
-        it,
-        op_name,
-        op_return_ntype,
-        ltoirs,
-    ):
-        self.it = it
-        self.ntype = op_return_ntype
-        self.prefix = f"transform_{it.prefix}_{op_name}"
-        self.ltoirs = ltoirs
-
-    @property
-    def state_c_void_p(self):
-        return self.it.state_c_void_p
+    @staticmethod
+    def dereference(it):
+        return it[0]
 
     @property
-    def size(self):
-        return self.it.size  # TODO fix for stateful op
+    def state(self):
+        return ctypes.cast(ctypes.pointer(self._cvalue), ctypes.c_void_p)
+
+
+class CountingIterator(IteratorBase):
+    def __init__(self, value):
+        value_type = numba.from_dtype(value.dtype)
+        self._cvalue = _ctypes_type_given_numba_type(value_type)(value)
+        numba_type = numba.types.CPointer(value_type)
+        abi_name = f"{self.__class__.__name__}_{str(value_type)}"
+        super().__init__(
+            numba_type=numba_type,
+            value_type=value_type,
+            abi_name=abi_name,
+        )
+
+    @staticmethod
+    def advance(it, distance):
+        it[0] += distance
+
+    @staticmethod
+    def dereference(it):
+        return it[0]
 
     @property
-    def alignment(self):
-        return self.it.alignment  # TODO fix for stateful op
-
-
-def TransformIterator(op, it):
-    # TODO(rwgk): Resolve issue #3064
-
-    if hasattr(it, "dtype"):
-        assert not hasattr(it, "ntype")
-        it = pointer(it, getattr(numba.types, str(it.dtype)))
-    it_ntype_ir = _ir_type_given_numba_type(it.ntype)
-    if it_ntype_ir is None:
-        raise RuntimeError(f"Unsupported: {type(it.ntype)=}")
-
-    op_abi_name = f"{op.__name__}_{it.ntype.name}"
-    op_ltoir, op_return_ntype = _ncc(op_abi_name, op, (it.ntype,))
-    op_return_ntype_ir = _ir_type_given_numba_type(op_return_ntype)
-
-    def source_advance(it_state_ptr, diff):
-        pass
-
-    def make_advance_codegen(name):
-        def codegen(context, builder, sig, args):
-            state_ptr, dist = args
-            fnty = ir.FunctionType(
-                ir.VoidType(), (_CHAR_PTR_IR_TYPE, _DISTANCE_IR_TYPE)
+    def state(self):
+        return ctypes.cast(ctypes.pointer(self._cvalue), ctypes.c_void_p)
+
+
+def make_transform_iterator(it, op):
+    if hasattr(it, "__cuda_array_interface__"):
+        it = pointer(it, numba.from_dtype(it.dtype))
+
+    it_advance = numba.cuda.jit(type(it).advance, device=True)
+    it_dereference = numba.cuda.jit(type(it).dereference, device=True)
+    op = numba.cuda.jit(op, device=True)
+
+    class TransformIterator(IteratorBase):
+        def __init__(self, it, op):
+            self._it = it
+            numba_type = it.numba_type
+            op_abi_name = f"{self.__class__.__name__}_{op.py_func.__name__}"
```

Comment:
Is this where we'll need to make the name more unique if there are two transform iterators in the same context in the future? Maybe add a TODO?

Reply:
Yup - added a TODO